README.md
# flutter_user_auth

A Flutter package to support user authentication through endpoint and also provide login and signUp ui

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [0.0.1]

Initial Release


lib/widgets/Logo.dart
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';

class Logo extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return RichText(
      textAlign: TextAlign.center,
      text: TextSpan(
          text: 'p',
          style: GoogleFonts.portLligatSans(
            textStyle: Theme.of(context).textTheme.display1,
            fontSize: 30,
            fontWeight: FontWeight.w700,
            color: Color(0xffe46b10),
          ),
          children: [
            TextSpan(
              text: 'iz',
              style: TextStyle(color: Colors.black, fontSize: 30),
            ),
            TextSpan(
              text: 'on',
              style: TextStyle(color: Color(0xffe46b10), fontSize: 30),
            ),
          ]),
    );
  }
}

lib/screens/SignUp.dart
import 'package:flutter_user_auth/flutter_user_auth.dart';
import 'package:flutter/material.dart';
import 'package:toast/toast.dart';
import 'package:http/http.dart' as https;
import 'package:dart_casing/dart_casing.dart';
import 'dart:convert';
import 'package:flutter_user_auth/widgets/BackButton.dart';
import 'package:flutter_user_auth/widgets/EntryField.dart';
import 'package:flutter_user_auth/widgets/Logo.dart';
import 'package:flutter_user_auth/widgets/SubmitButton.dart';
import 'package:flutter_user_auth/screens/Login.dart';
import 'package:flutter_user_auth/res/values/Strings.dart';

class SignUp extends StatefulWidget {
  @override
  _SignUpState createState() => _SignUpState();
}

class _SignUpState extends State<SignUp> {
  static String nameTitle = "Name";
  static String mobileNumberTitle = "Mobile Number";
  static String emailIdTitle = "Email Id";
  static String passwordTitle = "Password";

  String signUpMessage = Strings.EMPTY_STRING;

  final _formKey = GlobalKey<FormState>();
  var inputs = Map<String, String>();
  bool isProcessing = false;

  void _registerWithEmail() async {
    setState(() {
      signUpMessage = Strings.EMPTY_STRING;
    });
    FocusScope.of(context).unfocus();
    bool isValid = _formKey.currentState.validate();
    if (!isValid) return;

    _formKey.currentState.save();
    String nameInput = inputs[nameTitle];
    String emailIdInput = inputs[emailIdTitle];
    String mobileNumberInput = inputs[mobileNumberTitle];
    String passwordInput = inputs[passwordTitle];

    setState(() {
      isProcessing = true;
    });
    var response = await https.post(UserAuth.authConfig.registerConfig.endPoint,
        body: jsonEncode({
          "userType": UserAuth.authConfig.registerConfig.userType,
          Casing.camelCase(nameTitle): nameInput,
          Casing.camelCase(emailIdTitle): emailIdInput,
          Casing.camelCase(mobileNumberTitle): mobileNumberInput,
          Casing.camelCase(passwordTitle): passwordInput
        }));
    Map<String, dynamic> responseBody = json.decode(response.body);
    if (response.statusCode == 200) {
      await UserAuth.saveAuthToken(responseBody[Strings.AUTH_TOKEN]);
      Toast.show(Strings.REGISTERED_MESSAGE, context);
      Navigator.pop(context);
      return;
    }
    String message = responseBody[Strings.MESSAGE];
    if (response.statusCode == 400) {
      setState(() {
        isProcessing = false;
        signUpMessage = message;
      });
      return;
    }
    Navigator.pop(context);
    Toast.show(message, context);
  }

  void _onSave(String title, String input) {
    inputs[title] = input;
  }

  Widget _loginAccountLabel() {
    return Container(
      margin: EdgeInsets.symmetric(vertical: 20),
      alignment: Alignment.bottomCenter,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Text(
            'Already have an account ?',
            style: TextStyle(fontSize: 13, fontWeight: FontWeight.w600),
          ),
          SizedBox(
            width: 10,
          ),
          InkWell(
            onTap: () {
              Navigator.pop(context);
              Navigator.push(
                  context, MaterialPageRoute(builder: (context) => Login()));
            },
            child: Text(
              'Login',
              style: TextStyle(
                  color: Color(0xfff79c4f),
                  fontSize: 13,
                  fontWeight: FontWeight.w600),
            ),
          )
        ],
      ),
    );
  }

  // TODO: Take input from UserAuth RegisterConfig
  Widget _emailPasswordWidget() {
    return Column(
      children: <Widget>[
        EntryFieldWidget(
          title: nameTitle,
          validationFunction: EntryFieldValidator.emptyValidator,
          onSave: _onSave,
        ),
        EntryFieldWidget(
          title: mobileNumberTitle,
          isNumber: true,
          validationFunction: EntryFieldValidator.mobileNumberValidator,
          onSave: _onSave,
        ),
        EntryFieldWidget(
          title: emailIdTitle,
          validationFunction: EntryFieldValidator.emailIdValidator,
          onSave: _onSave,
        ),
        EntryFieldWidget(
          title: passwordTitle,
          isPassword: true,
          validationFunction: EntryFieldValidator.passwordValidator,
          onSave: _onSave,
        ),
      ],
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: isProcessing
            ? Center(
                child: CircularProgressIndicator(),
              )
            : SingleChildScrollView(
                child: Form(
                key: _formKey,
                onChanged: (){
                  setState(() {
                    signUpMessage = Strings.EMPTY_STRING;
                  });
                },
                child: Container(
                  height: MediaQuery.of(context).size.height,
                  child: Stack(
                    children: <Widget>[
                      Container(
                        padding: EdgeInsets.symmetric(horizontal: 20),
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.center,
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: <Widget>[
                            Expanded(
                              flex: 3,
                              child: SizedBox(),
                            ),
                            Logo(),
                            SizedBox(
                              height: 50,
                            ),
                            _emailPasswordWidget(),
                            SizedBox(
                              height: 50,
                              child: signUpMessage.isNotEmpty
                                  ? Center(
                                      child: Text(
                                        signUpMessage,
                                        style: TextStyle(
                                          color: Colors.red,
                                          fontSize: 15,
                                        ),
                                      ),
                                    )
                                  : null,
                            ),
                            SubmitButtonWidget(
                              onTap: _registerWithEmail,
                              title: 'Register Now',
                            ),
                            Expanded(
                              flex: 2,
                              child: SizedBox(),
                            )
                          ],
                        ),
                      ),
                      Align(
                        alignment: Alignment.bottomCenter,
                        child: _loginAccountLabel(),
                      ),
                      Positioned(top: 40, left: 0, child: BackButtonWidget()),
                    ],
                  ),
                ),
              )));
  }
}


lib/flutter_user_auth.dart
library flutter_user_auth;

import 'package:flutter/material.dart';
import 'screens/Login.dart';
import 'screens/SignUp.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'res/values/Strings.dart';

class UserAuth {
  static AuthConfig authConfig;
  static bool isLogged = false;

  static void initialize (AuthConfig authConfig) async{
    UserAuth.authConfig = authConfig;
    String authToken = await getAuthToken();
    if(authToken == null)
      isLogged = false;
    else
      isLogged = true;
  }

  static void login(BuildContext context) {
    Navigator.push(context, MaterialPageRoute(builder: (context) => Login()));
  }

  static void signUp(BuildContext context) {
    Navigator.push(context, MaterialPageRoute(builder: (context) => SignUp()));
  }

  static void logout() async{
    SharedPreferences prefs = await SharedPreferences.getInstance();
    await prefs.remove(Strings.AUTH_TOKEN);
    isLogged = false;
  }

  static Future<String> getAuthToken() async {
    SharedPreferences prefs = await SharedPreferences.getInstance();
    String authToken = prefs.getString(Strings.AUTH_TOKEN);
    return authToken;
  }

  static Future<void> saveAuthToken(String authToken) async{
    SharedPreferences prefs = await SharedPreferences.getInstance();
    await prefs.setString(Strings.AUTH_TOKEN, authToken);
    isLogged = true;
  }
}

class LoginConfig {
  String endPoint;
  var params = <String>[];

  LoginConfig({
    @required this.endPoint,
    this.params = const ['Email Id', 'Password'],
  });
}

class RegisterConfig {
  String endPoint;
  String userType;
  var params = <String>[];

  RegisterConfig({
    @required this.endPoint,
    @required this.userType,
    this.params = const ['Name, Mobile Number, Email Id, Password'],
  });
}

class AuthConfig {
  LoginConfig loginConfig;
  RegisterConfig registerConfig;
  bool saveUser;
  Widget logo;

  AuthConfig({
    @required this.loginConfig,
    @required this.registerConfig,
    this.saveUser = true,
    this.logo,
  });
}


lib/widgets/EntryField.dart
import 'package:flutter/material.dart';
import 'package:email_validator/email_validator.dart';

typedef String Validator(String text);
typedef void ValidityUpdater(String text, bool isValid);
typedef void SaverFunction(String title, String input);

class EntryFieldWidget extends StatefulWidget {
  EntryFieldWidget({
    @required this.title,
    this.isPassword = false,
    this.isNumber = false,
    this.validationFunction = EntryFieldValidator.noValidator,
    this.onSave,
  });
  final String title;
  final bool isPassword;
  final bool isNumber;
  final Validator validationFunction;
  final SaverFunction onSave;
  @override
  _EntryFieldWidgetState createState() =>
      _EntryFieldWidgetState(obscureText: isPassword);
}

class _EntryFieldWidgetState extends State<EntryFieldWidget> {
  _EntryFieldWidgetState({
    this.obscureText,
  });
  bool obscureText;
  FocusNode _focusNode = FocusNode();
  GlobalKey<FormFieldState> key = new GlobalKey<FormFieldState>();

  @override
  void initState() {
    super.initState();
    _focusNode.addListener(_focusNodeListener);
  }

  void _focusNodeListener() {
    if (!_focusNode.hasFocus) {
      key.currentState.validate();
    }
  }

  @override
  void dispose() {
    _focusNode.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: 90,
      child: TextFormField(
        key: key,
        keyboardType:
            widget.isNumber ? TextInputType.numberWithOptions() : null,
        focusNode: _focusNode,
        obscureText: obscureText,
        textInputAction: TextInputAction.next,
        validator: widget.validationFunction,
        onSaved: (String input){
          widget.onSave(widget.title, input);
        },
        decoration: InputDecoration(
          labelText: widget.title,
          border: InputBorder.none,
          errorBorder:
              OutlineInputBorder(borderSide: BorderSide(color: Colors.red)),
          enabledBorder: OutlineInputBorder(
              borderSide: BorderSide(color: Color(0xfff3f3f4))),
          focusedBorder:
              OutlineInputBorder(borderSide: BorderSide(color: Colors.blue)),
          focusedErrorBorder:
              OutlineInputBorder(borderSide: BorderSide(color: Colors.blue)),
          fillColor: Color(0xfff3f3f4),
          filled: true,
          suffixIcon: widget.isPassword
              ? IconButton(
                  icon: Icon(Icons.remove_red_eye),
                  onPressed: () {
                    setState(() {
                      obscureText = !obscureText;
                    });
                  })
              : null,
        ),
        onFieldSubmitted: (String text) {
          bool isValid = key.currentState.validate();
          if (isValid) FocusScope.of(context).nextFocus();
        },
      ),
    );
  }
}

class EntryFieldValidator {
  static String emailIdValidator(String email) {
    bool valid = EmailValidator.validate(email);
    if (valid) return null;
    return "Please enter valid Email ID";
  }

  static String mobileNumberValidator(String phone) {
    if (int.tryParse(phone) != null && phone.length == 10) return null;
    return "Please enter valid Mobile number";
  }

  static String emptyValidator(String text) {
    if (text.isNotEmpty) return null;
    return "Required Field";
  }

  static String noValidator(String text) {
    return null;
  }

  static String passwordValidator(String password) {
    if (password.length < 8) {
      return "Password should be 8 characters long";
    }
    return null;
  }
}


lib/widgets/SubmitButton.dart
import 'package:flutter/material.dart';

typedef void SubmitFunction();

class SubmitButtonWidget extends StatelessWidget {
  SubmitButtonWidget({
    @required this.onTap,
    this.title = "Default Title",
  });
  final SubmitFunction onTap;
  final String title;

  @override
  Widget build(BuildContext context) {
    return Material(
      type: MaterialType.transparency,
      child: Ink(
        width: MediaQuery.of(context).size.width,
        height: 50,
        decoration: BoxDecoration(
          borderRadius: BorderRadius.all(Radius.circular(5)),
          boxShadow: <BoxShadow>[
            BoxShadow(
                color: Colors.grey.shade200,
                offset: Offset(2, 4),
                blurRadius: 5,
                spreadRadius: 2)
          ],
          gradient: LinearGradient(
            begin: Alignment.centerLeft,
            end: Alignment.centerRight,
            colors: [Color(0xfffbb448), Color(0xfff7892b)],
          ),
        ),
        child: InkWell(
          onTap: onTap,
          child: Center(
            child: Text(
              title,
              style: TextStyle(fontSize: 20, color: Colors.white),
            ),
          ),
        ),
      ),
    );
  }
}

lib/screens/Login.dart
import 'package:flutter/material.dart';
import 'package:toast/toast.dart';
import 'package:http/http.dart' as https;
import 'package:dart_casing/dart_casing.dart';
import 'dart:convert';
import 'package:flutter_user_auth/flutter_user_auth.dart';
import 'package:flutter_user_auth/widgets/BackButton.dart';
import 'package:flutter_user_auth/widgets/EntryField.dart';
import 'package:flutter_user_auth/widgets/Logo.dart';
import 'package:flutter_user_auth/widgets/SubmitButton.dart';
import 'package:flutter_user_auth/screens/SignUp.dart';
import 'package:flutter_user_auth/res/values/Strings.dart';

class Login extends StatefulWidget {
  @override
  _LoginState createState() => _LoginState();
}

class _LoginState extends State<Login> {
  static String emailIdTitle = "Email Id";
  static String passwordTitle = "Password";

  String loginMessage = Strings.EMPTY_STRING;
  final _formKey = GlobalKey<FormState>();
  var inputs = new Map<String, String>();
  bool _isProcessing = false;

  void _loginWithEmail() async {
    setState(() {
      loginMessage = Strings.EMPTY_STRING;
    });
    FocusScope.of(context).unfocus();
    bool isValid = _formKey.currentState.validate();
    if (!isValid) return;

    _formKey.currentState.save();
    String emailIdInput = inputs[emailIdTitle];
    String passwordInput = inputs[passwordTitle];

    setState(() {
      _isProcessing = true;
    });
    var response = await https.post(UserAuth.authConfig.loginConfig.endPoint,
        body: jsonEncode({
          Casing.camelCase(emailIdTitle): emailIdInput,
          Casing.camelCase(passwordTitle): passwordInput
        }));
    Map<String, dynamic> responseBody = jsonDecode(response.body);
    if (response.statusCode == 200) {
      await UserAuth.saveAuthToken(responseBody[Strings.AUTH_TOKEN]);
      Toast.show(Strings.LOGGED_IN_MESSAGE, context);
      Navigator.pop(context);
      return;
    }
    final String message = responseBody[Strings.MESSAGE];
    if (response.statusCode == 400) {
      setState(() {
        _isProcessing = false;
        loginMessage = message;
      });
      return;
    }
    Navigator.pop(context);
    Toast.show(message, context);
  }

  void _onSave(String title, String input) {
    inputs[title] = input;
  }

  Widget _divider() {
    return Container(
      margin: EdgeInsets.symmetric(vertical: 10),
      child: Row(
        children: <Widget>[
          SizedBox(
            width: 20,
          ),
          Expanded(
            child: Padding(
              padding: EdgeInsets.symmetric(horizontal: 10),
              child: Divider(
                thickness: 1,
              ),
            ),
          ),
          Text('or'),
          Expanded(
            child: Padding(
              padding: EdgeInsets.symmetric(horizontal: 10),
              child: Divider(
                thickness: 1,
              ),
            ),
          ),
          SizedBox(
            width: 20,
          ),
        ],
      ),
    );
  }

  // TODO: Make Facebook button rounded and add Google button
  Widget _facebookButton() {
    return Container(
      height: 50,
      margin: EdgeInsets.symmetric(vertical: 20),
      decoration: BoxDecoration(
        borderRadius: BorderRadius.all(Radius.circular(10)),
      ),
      child: Row(
        children: <Widget>[
          Expanded(
            flex: 1,
            child: Container(
              decoration: BoxDecoration(
                color: Color(0xff1959a9),
                borderRadius: BorderRadius.only(
                    bottomLeft: Radius.circular(5),
                    topLeft: Radius.circular(5)),
              ),
              alignment: Alignment.center,
              child: Text('f',
                  style: TextStyle(
                      color: Colors.white,
                      fontSize: 25,
                      fontWeight: FontWeight.w400)),
            ),
          ),
          Expanded(
            flex: 5,
            child: Container(
              decoration: BoxDecoration(
                color: Color(0xff2872ba),
                borderRadius: BorderRadius.only(
                    bottomRight: Radius.circular(5),
                    topRight: Radius.circular(5)),
              ),
              alignment: Alignment.center,
              child: Text('Log in with Facebook',
                  style: TextStyle(
                      color: Colors.white,
                      fontSize: 18,
                      fontWeight: FontWeight.w400)),
            ),
          ),
        ],
      ),
    );
  }

  Widget _createAccountLabel() {
    return Container(
      margin: EdgeInsets.symmetric(vertical: 20),
      alignment: Alignment.bottomCenter,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Text(
            'Don\'t have an account ?',
            style: TextStyle(fontSize: 13, fontWeight: FontWeight.w600),
          ),
          SizedBox(
            width: 10,
          ),
          InkWell(
            onTap: () {
              Navigator.pop(context);
              Navigator.push(
                  context, MaterialPageRoute(builder: (context) => SignUp()));
            },
            child: Text(
              'Register',
              style: TextStyle(
                  color: Color(0xfff79c4f),
                  fontSize: 13,
                  fontWeight: FontWeight.w600),
            ),
          )
        ],
      ),
    );
  }

  // TODO: Take input from UserAuth LoginConfig
  Widget _emailPasswordWidget() {
    return Column(
      children: <Widget>[
        EntryFieldWidget(
          title: emailIdTitle,
          validationFunction: EntryFieldValidator.emailIdValidator,
          onSave: _onSave,
        ),
        EntryFieldWidget(
          title: passwordTitle,
          isPassword: true,
          validationFunction: EntryFieldValidator.passwordValidator,
          onSave: _onSave,
        ),
      ],
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: _isProcessing
          ? Center(
              child: CircularProgressIndicator(),
            )
          : SingleChildScrollView(
              child: Form(
                key: _formKey,
                onChanged: (){
                  setState(() {
                    loginMessage = Strings.EMPTY_STRING;
                  });
                },
                child: Container(
                  height: MediaQuery.of(context).size.height,
                  child: Stack(
                    children: <Widget>[
                      Container(
                        padding: EdgeInsets.symmetric(horizontal: 20),
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.center,
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: <Widget>[
                            Expanded(
                              flex: 3,
                              child: SizedBox(),
                            ),
                            Logo(),
                            SizedBox(
                              height: 50,
                            ),
                            _emailPasswordWidget(),
                            SizedBox(
                              height: 50,
                              child: loginMessage.isNotEmpty
                                  ? Center(
                                      child: Text(
                                        loginMessage,
                                        style: TextStyle(
                                            color: Colors.red, fontSize: 15),
                                      ),
                                    )
                                  : null,
                            ),
                            SubmitButtonWidget(
                              onTap: _loginWithEmail,
                              title: 'Login',
                            ),
                            Container(
                              padding: EdgeInsets.symmetric(vertical: 10),
                              alignment: Alignment.centerRight,
                              child: Text('Forgot Password ?',
                                  style: TextStyle(
                                      fontSize: 14,
                                      fontWeight: FontWeight.w500)),
                            ),
                            _divider(),
                            _facebookButton(),
                            Expanded(
                              flex: 2,
                              child: SizedBox(),
                            ),
                          ],
                        ),
                      ),
                      Align(
                        alignment: Alignment.bottomCenter,
                        child: _createAccountLabel(),
                      ),
                      Positioned(top: 40, left: 0, child: BackButtonWidget()),
                    ],
                  ),
                ),
              ),
            ),
    );
  }
}
